test.py

```diff
@@ -1,61 +1,110 @@
+        if isinstance(node, ast.FunctionDef) and node.name == func_name:
+            return True
+    return False
+
+
+def import_sdp_mods(include=None, ignore=None):
+    mods = []
+    for m in sorted(list(pkgutil.iter_modules(sdp.__path__))):
+        mod_path = f"sdp/{m[1]}.py"
+        if include is not None and len(include) and not any(mod in mod_path for mod in include):
+            continue
+        if ignore is not None and len(ignore) and any(mod in mod_path for mod in ignore):
+            continue
+
+        if (
+            "sdp" in m[1]
+            and "naive" not in m[1]
+            and func_exists(mod_path, "sliding_dot_product")
+        ):
+            mod_name = f"sdp.{m[1]}"
+            mod = importlib.import_module(mod_name)
+            mods.append(mod)
+
+    return mods
+
+
 if __name__ == "__main__":
-    modules = [
-        pyfftw_sdp,
-        njit_sdp,
-        numpy_fft_sdp,
-        scipy_oaconvolve_sdp,
-    ]
-
-    n_iter = 4
-    p_min = 6
-    p_max = 28
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-niter", default=4, type=int)
+    parser.add_argument("-pmin", default=6, type=int)
+    parser.add_argument("-pmax", default=27, type=int)
+    parser.add_argument("-ignore", default=None, nargs='*')
+    parser.add_argument("include", default=None, nargs='*')
+    args = parser.parse_args()
+
+    modules = import_sdp_mods(args.include, args.ignore)
+
+    n_iter = args.niter
+    p_min = args.pmin
+    p_max = args.pmax
+
+    print(f"module,len_Q,len_T,n_iter,time", flush=True)
+
     start_timing = time.time()
     for mod in modules:
-        for i in range(p_min, p_max):
+        mod_name = mod.__name__[4:-4]
+        for i in range(p_min, p_max + 1):
             Q = np.random.rand(2**i)
             break_Q = False
-            for j in range(i, p_max):
+            for j in range(i, p_max + 1):
                 T = np.random.rand(2**j)
                 break_T = False
 
                 elapsed_times = []
                 for _ in range(n_iter):
+                    try:
+                        mod.setup(Q, T)
+                    except:
+                        pass
                     start = time.time()
                     mod.sliding_dot_product(Q, T)
                     diff = time.time() - start
                     if diff > 10.0:
                         break_T = True
-                        warnings.warn(
-                            f"SKIPPED: {mod.__name__},{len(Q)},{len(T)},{diff})"
-                        )
+                        warnings.warn(f"SKIPPED: {mod_name},{len(Q)},{len(T)},{diff})")
                         break
                     else:
                         elapsed_times.append(diff)
 
                 if break_T:
-                    if j == 2 * i:
+                    if j == i + 1:
                         break_Q = True
                     break
 
-                elapsed_times.remove(
-                    min(elapsed_times)
```

Comment:
Is this part of code removed because of having the function `setup` in each `sdp`'s module now?

Reply:
Correct!

Reply:
Having said that, we still have a problem in that the `njit` method still performs the best when `len(Q) == len(T)` but we'll discuss that later when we are done with these changes.